### ISSUE TITLE
Open URL from (only) history

### ISSUE THREAD

**Issue by mlochbaum** [The-Compiler/qutebrowser#1885]
Currently `:open` completes with quickmarks, bookmarks, and URLs from history. Often the first two get in the way of what I actually want to do, which is to open a URL which I remember visiting, but haven't bookmarked. This is a particular problem when I want to open a page that I closed a few tabs ago, since I have to scroll through two categories before looking for it.

There are two obvious interfaces to use:
1. A flag `--history` for `:open`
2. A new command `:open-history`

I lean towards the second since the completion is easier to implement, but it does seem a bit silly to make a whole new command for something that's basically supported.

There doesn't appear to be a completion model for just the history now. Is that right? It also seems like it might be possible to make a history model and then get the URL model by composing bookmarks, quickmarks, and history in some way. I'd be willing to work on that if there's no obvious reason why it's not possible.

**Comment by The-Compiler** [comment]
#1853 added keybindings (Ctrl-Tab/Ctrl-Shift-Tab by default) to jump from completion to completion - do you think that'd suffice?

**Comment by mlochbaum** [comment]
I know about those, but it's still two extra keypresses for something I do pretty often. Jumping also only scrolls to show the first element in history so I have to scroll down further to see my options.

**Comment by lahwaacz** [comment]
If you say that you remember the URL, but haven't bookmarked, you should be able to type the unique part of the URL or title to get all quickmarks and bookmarks out of the way. As a bonus, you won't have to scroll down any further, since it should be the only item in the list. Admittedly, #1654 is not in yet, so it might not work perfectly yet...

**Comment by mlochbaum** [comment]
Remembering that I visited a page is very different from remembering the URL.

**Comment by lahwaacz** [comment]
In that case, if you go only for "a page closed a few tabs ago", isn't completion for `:undo` (part of #32) the thing you really want?

**Comment by The-Compiler** [comment]
If we really need such a feature, I'd like to think about a way to generalize it, i.e. being able to say "I want to show `:open`, but only with completions for quickmarks" somehow. This should be fairly easy to implement once completions are composable (something @rcorre is looking at), but I'm not sure how the UI would look for something like this.

**Comment by mlochbaum** [comment]
@lahwaacz Well, not really. I might have closed the browser since visiting the page, for instance.

The thing I want to do is really "go back to a page that I visited at some point in the past." There are a few cases in which I want to do this and I gave the one which is the hardest right now.
* If I knew at the time that I wanted to revisit the page, I would bookmark or possibly quickmark it. No problems here.
* If I visited the page in the distant past, I have to go back and search in the history file. I would like for qutebrowser to handle this (performantly), but it's only tangentially relevant here.
* If it's still in the in-memory history but not that recent, I try to remember part of the URL or title and search with that. Having marks at the top of the completion window is still annoying here, but not a huge problem.
* If it's within the last 20 pages or so, I would like to just press a key and start scrolling down. Occasionally I want to go further back without filtering.
* Sometimes I just want to look at my recent history (to see how I got to this page, maybe). Having marks in the way adds a lot of friction here too.

I view "look at the history" as somewhat different than "open a page", much like "go to a quickmark" is different from either of those. It's good for `:open` to suggest pages that I might want to open, but it doesn't replace my desire to sometimes look at the history. It's also worth noting that if we ever implement `:open` history completion by frecency, then the two get even further apart.

**Comment by mlochbaum** [comment]
@The-Compiler I hadn't seen @rcorre's completion refactoring, but that does look like it helps a lot with this problem. After some thought about this I would probably campaign for a separate `:history` command anyway, but being able to control completions better would probably be acceptable.

**Comment by rcorre** [comment]
A few thoughts:

>  Jumping also only scrolls to show the first element in history so I have to scroll down further to see my options

I didn't think of this when I implemented it, but placing the category at the top makes more sense.

> I have to go back and search in the history file

How about a `qute://history` page with a similar format to `qute://log`? You could browse through your history in a nice tabular format and use hinting to navigate. In addition, it could display your whole history file rather than limiting it to 1000 as `open` does (by default).

> it does seem a bit silly to make a whole new command for something that's basically supported

This is basically what `bookmark-open` is. That doesn't make it a good idea, however :smile:

I actually _do_ plan on implementing flag-scoped completion. For example, `:bind --mode=caret` will only offer to complete commands that can be used in caret mode. However, `--mode=caret` actually has meaning for how the command executes.  A flag like `--history` that exists only to scope completion (and is ignored by the command itself) seems odd.

> composing bookmarks, quickmarks, and history

Yup, composable completion sections is part of the refactoring plan, but it is held back a bit by `COLUMN_WIDTHS` being explicitly specified on each model.

**Comment by mlochbaum** [comment]
`qute://history` is a good idea. I'll see if I can implement it. Of course it's a bit slower than a command, but it gives you some extra functionality like seeing the context for history after you search.

For the `COLUMN_WIDTHS` problem, just compose and then override `COLUMN_WIDTHS`? It seems like data that you couldn't possibly get from the composed models.

**Comment by The-Compiler** [comment]
For the record, I fully agree with the points @rcorre has made. :wink:

> If I visited the page in the distant past, I have to go back and search in the history file. I would like for qutebrowser to handle this (performantly), but it's only tangentially relevant here.

I still hope #1765 would help tremendously help with that (as Qt can simply lazily load more data when required)

**Comment by lahwaacz** [comment]
What about setting a fixed completion model _before_ the actual command? Basically, in python/qutebrowser pseudocode:

```
with :completion-model --history:
    :open
```

In qutebrowser, this could be written simply as `:completion-model --history -- :open`, but there has to be a state holding the completion model to be used by the next command. Bare `:completion-model` or plain commands would default to the current completer autochanging based on the current prompt string, as typing `:set section option value` invokes 4 different models.

**Comment by mlochbaum** [comment]
That can be part of `:set-cmd-text`. I was thinking about that idea after @rcorre's comment, but I do think it's a pretty awkward way of specifying the completion.

However, I definitely want a flag for `:set-cmd-text` that opens the completion window and highlights the first entry for the following command (and will probably submit a PR to that effect soon).

**Comment by rcorre** [comment]
> I still hope #1765 would help tremendously help with that

I haven't abandoned that idea :smile: 
I'll probably look into that out of curiosity if nothing else.

> However, I definitely want a flag for :set-cmd-text that opens the completion window and highlights the first entry for the following command

In theory, would `set-cmd-text <text> ;; completion-item-next` do what you want? It doesn't actually seem to do anything for me, but it seems like it should (maybe this is a bug or I just don't understand how `;;` works)

**Comment by lahwaacz** [comment]
> However, I definitely want a flag for `:set-cmd-text` that opens the completion window and highlights the first entry for the following command

I don't think this makes sense to merge with `:set-cmd-text`. I was thinking that it could be implemented by simply binding `set-cmd-text -s :open ;; completion-item-focus next`, but that does not work as expected. Maybe you'd need `completion-item-focus` to wait after the completion is fully initialized?

**Comment by The-Compiler** [comment]
Yeah - right now a lot happens async, and a command can't really signal "I'm done" in some way - so `completion-item-focus` will run right after `set-cmd-text` but before the completion had a chance to open.

I guess you could do `set-cmd-text -s :open ;; later 10 completion-item-focus next` or so for now :laughing:

**Comment by mlochbaum** [comment]
The idea is that for some commands I want to actually type something, and for others I just want to use one of the completion options. This is reasonably similar to `-s` for `set-cmd-text`, which basically differentiates between a command with provided arguments I might want to modify and a command with no arguments provided.

I realize now that I didn't really want to highlight the first item since that invalidates future typing to narrow down the options. What I really want is to not need an extra keypress to begin scrolling. `set-cmd-text -s :open ;; later 10 completion-item-focus next` does do what I said I wanted to do, however.

In related news, I've just discovered that `completion-item-focus` crashes if you give it an argument that's not one of the four expected ones.

**Comment by lahwaacz** [comment]
> It would be nice if `completion-item-focus` with no arguments highlighted the first item but didn't fill the command line.

If you highlight the first item this way, what would you expect to happen when you press `Enter` afterwards? Complete the command according to the selected item and run it, or run the command uncompleted? The former does not fit with the current completion behaviour and the latter does not fit with the visual representation of the completion (why was the selected item not entered?). Also, with the later you can't easily enter the first item into the prompt.

**Comment by The-Compiler** [comment]
FYI, I just fixed #1871 in 12061b8bb12e9b385d80e83b4ba7411639c666a6 :wink:
